Bl5321

kkamdin opened the pull request:
PR-ing David's updates to writeNX.py

davidalexandershapiro committed:
Complete rewrite of writeNX:
1. new internal data structure
2. much cleaner allocation of arrays
3. follows the NXstxm structure closely
4. cleaner dataHandler addDataToStack
5. minor adjustments to scanning routines to accommodate
6. tested in simulation mode
7. not tested with real hardware.  Likely fails for spiral scans
8. still need to adapt readNexus

davidalexandershapiro committed:
updated readNexus in writeNX to account for file type 3.

davidalexandershapiro committed:
small correction to scripter allMotorPositions

davidalexandershapiro committed:
Added single and double motor scan drivers and related functionality to the gui.
This is tested in simulation mode but not yet with instruments
changes to:
mainwindow: manage single motor scan plots and double motor scan images
controller: minor changes to read_daq
dataHandler: manage single pixel data, resolved problem with queue syncing
writeNX: add motor names for scanned motors

davidalexandershapiro committed:
tested single/double motor scan against hardware at 5321.  Now working.
tested derivedPiezo coarse_only scan against hardware at 5321.  Now working.
Minor fixes to mainwindow.py for single/double motor scan display

davidalexandershapiro committed:
Fixed up writeNX to be compliant as tested by nxvalidate.
Removed the Z position axis from entry0/data/data so that the output is properly read by AXIS2000 and Mantis
Revisions to scan routines to account for the new dimensionality: line_image, line_focus, line_spectrum
Still need revisions to the derived scans

davidalexandershapiro committed:
New drivers for PI E712, SmarAct MCS2 and Quantum Detectors Xpress3
Some additional scripts for testing these drivers

davidalexandershapiro committed:
added E712 test script

davidalexandershapiro committed:
Merge branch 'tender250331' into bl5321

davidalexandershapiro committed:
Fixed spiral_image and derived_spiral_image to work with new writeNX
Tested against the MCL driver live with hardware

davidalexandershapiro committed:
Some cleanup in the config files
Added pyqtdarktheme and a theme switching capability

kkamdin commented on README.md:
```diff
@@ -43,13 +43,14 @@ sudo apt install  libxcb-cursor0
 - python-dotenv
+- PIPython
 
 # Environment setup and installation using miniconda3 and pip
```
below in the section # Running pystxmcontrol I think there may need to be some updates

I had to do to the following to be able to start the server:
- min config/main.json:
--server['ip'] was already 127.0.0.1
-- edit server['data_dir'] to a local path
- `copy config/*.json to /path/to/python/env/pystxmcontrol_cfg/`
- not sure what "edit [path_to_conda_env]/pystxmcontrol_cfg/main.json" or "[path_to_conda_env]/pystxmcontrol_cfg/motorConfig.json" is referring to (also checked the docs)

davidalexandershapiro replied:
If you pip install without the -e option, then the config files are copied to the correct location (into the conda environment, at least that's where the code is looking for them).  With -e, they need to be manually copied.  I don't know how to copy the data files to the expected location while using -e, or maybe some other way of managing the files is needed.  It's a hassle I'd like to solve

The two bullet points under "edit main.json" explain what should be edited.  I'll make them sub-bullets so it's clearer.  For me, [path_to_conda_environment] is, for example, ~/miniconda3/envs/stxm.  Not sure how to make that clearer but I"m sure there's a way!

There are docs in docs/build/html that describe how to edit motorConfig.json for a particular system.  It's html but I don't know how to have github serve that up.  I know it's possible just haven't tried to do it yet

kkamdin replied:
ooooh yeah sub-bullets would help. My brain was totally not putting together that the second two bullets were referring to the file named in the first bullet. 

It's possible my .json files didn't copy to the right place because I'm not actually using conda. Since you say it's something you'd like to solve, I can take a peek at it to see if I can figure out a way to do the copy regardless of pip install -e or not.

kkamdin commented on config/main.json:
```diff
@@ -26,67 +31,82 @@
-                    "yStep": 1.0,
-                    "xRange": 70.0,
-                    "yRange": 70.0,
+                    "xStart": -12.475,
```
is there a good place in the repo the to add a note about where these numbers (and other numbers for the other scan types) are coming from? a comment in this file maybe? feel free to ignore if this is obvious for typical use.

davidalexandershapiro replied:
They're really just place holders.  For actual use, these numbers get overwritten by the software to reflect the most recently executed scans, then they are read on startup.  The values in the repo are just what show up on the very first use and are more or less arbitrary

kkamdin replied:
cool we can ignore this comment then

kkamdin commented on pystxmcontrol/utils/writeNX.py:
this is a comment for utils/alsapi.py (which is not in this PR).

I couldn't get this message to go away. I tried putting my .env in many different places:

```
Request is not authorized. Your API key might be missing. Create a file named '.env' in the same directory as this notebook. In that file, paste this text...
ALSHUB_API_KEY='your-api-key-goes-here-inside-the-quotes'

If that does not work, ask Padraic to add your IP address to the firewall settings.
```

davidalexandershapiro replied:
I can remove this.  It's printing a traceback after an exception.  But maybe you can recommend a better way to manage this.  The GUI is looking for the ALSHUB_API_KEY as a system environment variable.  It uses this to talk to ALSHUB and get the proposal data.  In general this won't exist for users outside of the ALS but will exist at ALS beamlines that have set this up.

kkamdin replied:
I don't think you need to remove the traceback. My comment was more that I made a .env file with a fake ALSHUB_API_KEY defined in it, just to see if I could get rid of the exception but `load_dotenv()` didn't seem to find it no matter where I put the .env file. I thought the exception was from not having the ALSHUB_API_KEY environment variable set as opposed to not having a 'good' value for the api key (maybe I was wrong about that?). 

Are you able to run this locally with a fake ALSHUB_API_KEY defined? and just see that the exception print out disappears?

davidalexandershapiro replied:
Oh, right, I read the traceback too quickly.  This actually comes from a failed API request because of a bad key. You could talk to Dylan about getting a functional one, I think it's pretty quick.

kkamdin replied:
oh okay cool, he's on travel this week so I'll follow up later. We can consider this question of mine resolved. 

I think at some point we were talking about having users actually log into the gui with their ALS hub credentials so we can populate data authorization fields, etc. automagically? that can be a problem for future us...

kkamdin commented on pystxmcontrol/controller/dataHandler.py:
```diff
@@ -79,9 +79,6 @@ def getScanName(self, dir = None, prefix = None, ptychography = False):
         self.currentScanID = os.path.join(scanDir, fileName)
```
up above at line 57:

either use os.makedirs() or note somewhere that the server['data_dir'] path in main.json has to exist. otherwise crash

davidalexandershapiro replied:
I've never used os.makedirs() but it looks like it will just make the directories if they don't exist.  I"m not sure that's super safe in this case.  Since we generate a lot of data which should be put in a specific location, the user needs to take care to put it in the right spot.  So manually setting up this directory should be part of the initial set up, I think.  But I don't know the best way to enforce that (or at least handle it) with code.  Maybe just better documentation as you say.  I made some changes to the README that should clear that up.

kkamdin replied:
sounds good!

kkamdin commented on pystxmcontrol/gui/mainwindow.py:
```diff
@@ -530,14 +575,11 @@ def mouseMoved(self, pos):
+        # self.consoleText.insert(0,message)
+        # printStr = ''
+        # for item in self.consoleText:
+        self.consoleStr += '['+str(message["time"])+']\n' + json.dumps(message) + '\n\n'
```
I'm getting a key error from this line:
```
Traceback (most recent call last):
  File "/Users/kkamdin/Development/pystxmcontrol/pystxmcontrol/gui/mainwindow.py", line 581, in printToConsole
    self.consoleStr += '['+str(message["time"])+']\n' + json.dumps(message) + '\n\n'
KeyError: 'time'
```

I haven't tracked down where the message is getting built yet but is it missing the time stamp?

davidalexandershapiro replied:
That's fixed in the latest.  The server receives a message, processes it, adds a timestamp and then sends that out in a response.  There was one type of request that wasn't adding a timestamp.  Now it is!

kkamdin reviewed (commented):
This is a very light review of the code just calling out the couple of issues I had when getting up and running.

I'll check out writeNX.py in more detail separately!

davidalexandershapiro committed:
Fixed some issues in ptychography_image and writeNX so that ptychography scans work now with the new NX structure
Cleaned up some things in mainwindow

kkamdin commented on pystxmcontrol/utils/writeNX.py:
```diff
@@ -153,37 +100,47 @@ def _rebinLine(self, xReq, yReq, xMeas, yMeas, rawCounts):
-            #If no version is in the file, assign it to version 2.
-            self.meta["version"] = 0
-        if self.meta["version"] == '2':
+        if self.meta["version"] == 2.0:
```
maybe worth it to write a short comment at each if statement to give these version numbers some context. e.g `# version 2.0 is most files from 2024. region is stored differently in newer versions` or whatever.

davidalexandershapiro replied:
Regarding the top comment about "ptychography_image", I forgot to change that in the repo version of the config so I will fix that.  Regarding "defocus", that is in the current version.  But this does bring up a point about how to properly manage the config files.  New items won't necessarily make it into an existing users config unless they overwrite their current files, which then deletes their motor/scan info.  So this currently has to be done manually.  Is there a better solution?

Regarding comments for the file versions, point taken!  I will add some comments but it looks like that piece of code is from the main branch

kkamdin replied:
re: is there is a better solution / partial re-writes of config files. I think there are some tricks we can play... let me think about it and I'll propose something.

kkamdin commented on pystxmcontrol/utils/writeNX.py:
```diff
@@ -153,37 +100,47 @@ def _rebinLine(self, xReq, yReq, xMeas, yMeas, rawCounts):
-        self.meta["scan_type"] = f["entry0/counter0/stxm_scan_type"][()][0].decode()
+        # try:
+        f = h5py.File(stxm_file,'r')
+        self.NXfile = f
```
assignment of a class variable outside of initialization makes code unpredictable / hard to debug because the class has different attributes depending on the code path taken.

kkamdin replied:
btw -- usually when I comment I try to also explicitly suggest a change, but in this case I'm not sure what the solution is since I don't know the usage pattern for this class.

davidalexandershapiro replied:
Point taken.  I"ll put that in init

kkamdin commented on pystxmcontrol/utils/writeNX.py:
```diff
@@ -418,97 +322,40 @@ def startOutput(self):
         print("Starting output of file %s" %self.file_name)
         self.NXfile = h5py.File(self.file_name,'w',libver='latest')
```
same class variable used for reading and writing, could get some weird behavior

kkamdin replied:
again -- I'm not familiar enough with the usage patterns to suggest an explicit solution. So this is currently just a note about a non-standard code pattern.

davidalexandershapiro replied:
Right, I guess this should just happen in init.  I don't remember now why I separated it out into a separate function.  I think this is because the process is as follows:
1. initialize stxm object with a scan dict which doesn't yet know about the file name
2. determine correct file name
3. start output which creates the file and populates the entry

I could instead do 2 first and put the filename into the dict which initializes the class, then it could automatically start output.  I'll work on that.

kkamdin replied:
cool -- just to make sure we're on the same page, this comment was more about having the same variable for reading and writing, so like re-defining self.NXfile to be a writer where it was previously defined to be a reader in the course of program execution. I'm unsure of what parts of the data hander / control loop is writing and what part is reading, so it's possible this re-definition never happens, but it's possible for someone using the class to accidentally do that.

If you're okay moving these to the init, I'd suggest having one variable be the writer pointer and one variable be the reader pointer (they are pointing to the same file). so something like:

```
self.NXfile_writer =  h5py.File(self.file_name,'w',libver='latest')
self.NXfile_writer.swrm_mode = True
self.NXfile_reader =  h5py.File(self.file_name,'r')
```

Again, you're the expert on if this makes sense in the wider context of things. If this solution doesn't make sense, let me know if you want to talk over other options, maybe easier in person.

davidalexandershapiro replied:
OK, now I see the point.  I can do that.  I think nothing outside of the class needs access to it so they (or it) could also be private.

kkamdin commented on pystxmcontrol/utils/writeNX.py:
```diff
@@ -418,97 +322,40 @@ def startOutput(self):
-                self.saveRegion(i)
+        for i in range(n_scan_regions):
+            self.saveRegion(i)
         self.NXfile.swmr_mode = True
```
I think you have to define the SWMR at the h5py.FIie level as in: `h5py.File('name', 'r', swmr=True)`. I'm not sure what this variable is doing because it's not used again in this file.

davidalexandershapiro replied:
You can do it this way according to the docs: https://docs.h5py.org/en/stable/swmr.html

kkamdin replied:
my bad thank you for the docs link!

kkamdin commented on pystxmcontrol/utils/writeNX.py:
```diff
@@ -418,97 +322,40 @@ def startOutput(self):
-                    motor_grp.create_dataset(key, data=self.motorPositions[i][key])
+            self.updateEntry(i)
+        self.end_time = datetime.datetime.now()
+        self.NXfile['entry%i/end_time' %i][...] = str(self.end_time).encode("UTF_8")
```
I'd recommend updating to the new string format syntax here and elsewhere: `[f'entry{i}/end_time']`

kkamdin replied:
I was also just wondering about the choice of utf8 encoding vs string formatting? Nexus date times are supposed to be iso8601, so I think here you may want:

`self.NXfile[f'entry{i}/end_time'][...] = self.end_time.isoformat()`

davidalexandershapiro replied:
I changed the string format syntax! It seems NX prefers UTF_8 for character encoding.  I"m no expert on these things, maybe h5py does that automatically for strings?  I also changed the times to isoformat.  This is tested on my machine but not yet committed.  I"ll finish all remaining and commit.

kkamdin commented on pystxmcontrol/utils/writeNX.py:
```diff
@@ -517,292 +364,70 @@ def addFrame(self,frame,framenum, mode="dark"):
+        ne,nz_m,ny_m,nx_m = len(self.energies),len(self.zMeasured[i]),len(self.yMeasured[i]),len(self.xMeasured[i])
+        nz_r,ny_r,nx_r = len(self.zPos[i]), len(self.yPos[i]), len(self.xPos[i])
+
+        nxentry = self.NXfile.create_group("entry%i" %i)
```
another opportunity for new string format:

`        nxentry = self.NXfile.create_group(f'entry{i}')
`

davidalexandershapiro replied:
Done for all in writeNX

kkamdin commented on pystxmcontrol/utils/writeNX.py:
```diff
@@ -517,292 +364,70 @@ def addFrame(self,frame,framenum, mode="dark"):
+
+        nxentry = self.NXfile.create_group("entry%i" %i)
+        nxentry.attrs["NX_class"] = np.bytes_("NXentry")
+        start_time = nxentry.create_dataset("start_time", data=str(self.start_time).encode("UTF_8"))
```
a few notes on this section:

Since you're not using the `start_time` variable again in this function you can skip the output value and just have the stuff after the equal sign which also is nice because `start_time` here is actually a hdf5 dataset and not a time, so it makes things a little clearer:

`nxentry.create_dataset("start_time", data=str(self.start_time).encode("UTF_8"))
`
Same comment for `end_time`, `title`, `version`, etc.


Again I'm not sure about the UTF-8 encoding. This may be just two ways to do the same thing, but I think this is the preferred syntax, and it lets hdf5 handle the encoding for you, which is more future-proof. 

`nxentry.create_dataset("start_time", data=self.start_time.isoformat(), dtype=h5py.string_dtype())`

I think even for other strings you want to store you can do e.g.:
`nxentry.create_dataset("title", data=self.scan_dict["proposal"], dtype=hf5py.string_dtype())`

davidalexandershapiro replied:
I removed the unused values.  I"m uncertain about the string formatting.  NX documentation requests UTF-8.  Does h5py use that by default?  According to gemini: h5py.string_dtype(encoding='utf-8', length=None)

kkamdin replied:
yes, the `dtype=h5py.string_dtype()` will automatically use hdf5's preferred encoding (utf-8) -- I tested this method locally and it works just fine with the `dtype` keyword. The benefit is that you're letting the h5py package handle its own encoding, which means less code for you to maintain. 

Using `h5py.string_dtype(encoding='utf-8', length=None)` is similar to using encode("UTF-8"). It does use the h5py package to do the encoding, but it's still it's a lower level instruction where you are telling h5py what encoding to use.

I did a little more googling and this encoding handling is new for hdf5 > v2.10 -- apparently h5py fully handles its own encoding now so even this should work:

`nxentry.create_dataset("title", data=self.scan_dict["proposal"])`

kkamdin commented on pystxmcontrol/utils/writeNX.py:
```diff
@@ -517,292 +364,70 @@ def addFrame(self,frame,framenum, mode="dark"):
+        end_time = nxentry.create_dataset("end_time", data=str(self.end_time).encode("UTF_8"))
+        title = nxentry.create_dataset("title", data=self.scan_dict["proposal"].encode("UTF_8"))
+        version = nxentry.create_dataset("version", data=float(self.scan_dict["main_config"]["server"]["nx_file_version"]))
+        definition = nxentry.create_dataset("definition", data=["NXstxm".encode("UTF_8")])
```
I think this is required to be a single value by NXstxm? Are you getting an error if it's not a list?

`nxentry.create_dataset("definition", data="NXstxm", dtype=h5py.string_dtype())
`

davidalexandershapiro replied:
One of the open source softwares for stxm data wants it to be a one element long list.  I don't think it should be.  I don't think that code is actively developed but I know the developer and could ask

kkamdin replied:
Oh okay I remember this is what you demo-ed to me a few weeks ago. I think we mostly decided to go with NXstxm official format and encourage others to follow?

kkamdin reviewed (commented):
Okey dokey. I did a short review focusing on writeNX.py. I tried to keep the suggestions isolated to small changes; the stxm class mixes a lot of responsibilities in a way that I think makes the code difficult to maintain but didn't want to try to suggest a refactor right now, especially not really understanding myself how dataHander.py and other places use it. If that's something we think we want to tackle now we should chat more about what I think that may look like!

Also, I'm still getting an error when I try to run a ptychography scan, here's the full error thread: This should probably be updated in `main.json` and `scans.json`? that's the only place i still see `ptychography_grid`

```
  File "/Users/kkamdin/.local/share/uv/python/cpython-3.10.17-macos-aarch64-none/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/Users/kkamdin/.local/share/uv/python/cpython-3.10.17-macos-aarch64-none/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/kkamdin/Development/pystxmcontrol/pystxmcontrol/controller/controller.py", line 197, in scan_helper
    eval(scan["driver"]+"(scan, self.dataHandler, self, self.scanQueue)") #image_scan(scan)
  File "<string>", line 1, in <module>
NameError: name 'ptychography_grid' is not defined. Did you mean: 'ptychography_image'?
```

When I updated my local .json files so the driver is called `ptychography_image` I got the scan to run, but I ran into another issue, which may just be user error? Here's the output from that:

```
Starting output of file /Users/kkamdin/Development/pystxmcontrol/data/2025/06/250626/NS_250626004.stxm
Started data process.
starting ptychography scan:  /Users/kkamdin/Development/pystxmcontrol/data/2025/06/250626/NS_250626004
Done!
refocusing
Starting output of file /Users/kkamdin/Development/pystxmcontrol/data/2025/06/250626/NS_250626004_ccdframes_0_0.stxm
Exception in thread Thread-8 (scan_helper):
Traceback (most recent call last):
  File "/Users/kkamdin/.local/share/uv/python/cpython-3.10.17-macos-aarch64-none/lib/python3.10/threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "/Users/kkamdin/.local/share/uv/python/cpython-3.10.17-macos-aarch64-none/lib/python3.10/threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/kkamdin/Development/pystxmcontrol/pystxmcontrol/controller/controller.py", line 197, in scan_helper
    eval(scan["driver"]+"(scan, self.dataHandler, self, self.scanQueue)") #image_scan(scan)
  File "<string>", line 1, in <module>
  File "/Users/kkamdin/Development/pystxmcontrol/pystxmcontrol/controller/scans/ptychography_image.py", line 190, in ptychography_image
    step = energies[0] / 700. * controller.main_config["ptychography"]["defocus"]#15.
KeyError: 'defocus'
```

kkamdin commented on pystxmcontrol/utils/writeNX.py:
```diff
@@ -153,37 +100,47 @@ def _rebinLine(self, xReq, yReq, xMeas, yMeas, rawCounts):
+        #     return
         self.nRegions = len(list(f))
         self.data = {}
+        self.meta = {}
```
same comment for these class variables / assigned outside of the initialization. These variables won't exist unless we happen to always readNexus() before we try to access them.

davidalexandershapiro committed:
moved rebinLine from writeNX into utils/general
in writeNX it's now looking for the file_name in scan_dict when init is called.  the file name is now put in the dict by dataHandler
If there's no file_name then it just sets up the data structure but doesn't create a file

davidalexandershapiro committed:
revised writeNX to have separate reader/writer attributes.  Now called _nx_reader and _nx_writer

davidalexandershapiro committed:
Revised how single_motor_scan and mainwindow handle single motor scans where energy is the scanned axis.  It is a special
degree of freedom and now handled fully separately.

davidalexandershapiro committed:
added load_config function to the gui that reloads the config from the server
